### PR TITLE
make: boards: refactor board makefiles

### PR DESCRIPTION
**Pull request by kaspar030**
They were *mostly* the same, so now they include common parts.
This is includes #2850.

Just looked at Cortex-M boards, excluding again the kinetis boards.

The changes are a little less trivial, but if "hello world" flashes and displays using ```make flash``` and ```make term```, everything is fine.

Tested so far on:

- [x] airfy-beacon (nrf51822)
- [x] arduino-due (sam3x8e)
- [x] cc2538dk (cc2538)
- [x] f4vi1 (stm32f4)
- [x] fox (stm32f1)
- [x] iot-lab_M3 (stm32f1)
- [x] mbed_lpc1768 (lpc1768)
- [x] nucleo-f091 (stm32f0)
- [x] nucleo-f334 (stm32f3)
- [x] pca10000 (nrf51822)
- [x] pca10005 (nrf51822)
- [x] spark-core (stm32f1)
- [x] stm32f3discovery (stm32f3)
- [x] stm32f4discovery (stm32f4)
- [x] yunjia-nrf51822 (nrf51822)
- [x] msbiot (stm32f4)
- [x] nucleo-l1 (stm32l1)
- [x] openmote (cc2538)
- [x] samr21-xpro (samd21)
- [x] stm32f0discovery (stm32f0)
- [x] udoo (sam3x8e)

### CODE REVIEW

**Comment by kaspar030** [comment]
btw, this implicitly includes #2848.

File: sys/newlib/syscalls.c [hunk 1]
```diff
@@ -7,15 +8,16 @@
  */
 
 /**
- * @ingroup     cpu_stm32f4
+ * @ingroup     sys_newlib
  * @{
  *
  * @file
- * @brief       NewLib system call implementations for STM32F4
+ * @brief       NewLib system call implementations
```
Comment (Kijewski) [inline sys/newlib/syscalls.c]: `Newlib` calls itself `Newlib` (lowercase L)

File: sys/newlib/syscalls.c [hunk 2]
```diff
@@ -306,7 +309,12 @@ int _stat_r(struct _reent *r, char *name, struct stat *st)
 int _isatty_r(struct _reent *r, int fd)
 {
     r->_errno = 0;
-    return -1;
+
+    if(fd == STDOUT_FILENO || fd == STDERR_FILENO) {
```
Comment (Kijewski) [inline sys/newlib/syscalls.c]: + STDIN_FILENO

File: sys/newlib/Makefile.include [hunk 3]
```diff
@@ -0,0 +1 @@
+UNDEF := $(BINDIR)newlib/syscalls.o $(UNDEF)
```
Comment (Kijewski) [inline sys/newlib/Makefile.include]: Why not `+= UNDEF := $(BINDIR)newlib/syscalls.o`?

Comment (kaspar030) [inline sys/newlib/Makefile.include]: I wanted to change as little as possible to the original makefiles, and they all specify syscalls.o before startup.o. So here I prepend syscalls.

File: sys/newlib/syscalls.c (hunk 2 above)
Comment (kaspar030) [inline sys/newlib/syscalls.c]: will change. That was taken from a random old implementation.

File: sys/newlib/syscalls.c (hunk 1 above)
Comment (kaspar030) [inline sys/newlib/syscalls.c]: yup

**Comment by kaspar030** [comment]
arduino due fixed, works now

**Comment by kaspar030** [comment]
addressed @Kijewski's comments

**Comment by kaspar030** [comment]
should I split the cpu + board specific commits in seperate PR's? I'm afraid we'll never test all these boards...

**Comment by haukepetersen** [comment]
no, lets leave it in one PR, its fast to test (I will start now to test every board I have flying around here...)

**Comment by haukepetersen** [comment]
samr21-xpro does not build.

**Comment by haukepetersen** [comment]
udoo seems to be broken.

**Comment by kaspar030** [comment]
SAMR21 should be building now. I had left that out before because of the gpio issues caused by -Os.

**Comment by kaspar030** [comment]
> udoo seems to be broken.

Hm, I don't have a board...

**Comment by kaspar030** [comment]
Does anyone have these boards and can test?
- udoo (seems broken, can't fix without board)
- openmote
- stm32f0discovery
- msbiot
- cc2538dk
- samr21-xpro
- nucleo-l1

**Comment by haukepetersen** [comment]
I have all besides the `cc2538dk` and the `nucleo-l1` available somewhere. But can't promise when I will get to testing them.

**Comment by kaspar030** [comment]
- rebased

**Comment by OlegHahm** [comment]
What's the state of this PR?

**Comment by kaspar030** [comment]
Needs testing for these boards:

Hauke will give me this soon:
- udoo

Hauke has them around but doesn't find time:
- openmote
- stm32f0discovery
- msbiot
- samr21-xpro

Don't know who has them:
- cc2538dk
- nucleo-l1

If "make flash term" for hello-world works, the test is basically complete.

**Comment by OlegHahm** [comment]
I have a nucleo l1, but not sure if I have in the office in Paris. Will check on Monday.

Can test samr21-xpro.

**Comment by OlegHahm** [comment]
Maybe @hexluthor can help for cc2538dk?

**Comment by kaspar030** [comment]
Nice. samr21-xpro might stumble over the gpio implementation bug...

**Comment by OlegHahm** [comment]
So, basically we are waiting for a fix for the SAMR21 GPIO bug?

**Comment by hexluthor** [comment]
`make BOARD=cc2538dk -C examples/hello-world flash term` works for me on cc2538dk. OpenMote will probably work too, but I don't have one to test with.

**Comment by LudwigOrtmann** [comment]
```
make clean all 
Building application "hello-world" for "samr21-xpro" with MCU "samd21".

"make" -C /home/lo/RIOT/boards/samr21-xpro
"make" -C /home/lo/RIOT/core
"make" -C /home/lo/RIOT/cpu/samd21
"make" -C /home/lo/RIOT/cpu/cortex-m0_common
"make" -C /home/lo/RIOT/cpu/samd21/periph
/home/lo/RIOT/cpu/samd21/periph/gpio.c: In function 'isr_eic':
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1075:24: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[9].cb(gpio_config[9].arg);
                        ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1075:42: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[9].cb(gpio_config[9].arg);
                                          ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1079:24: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[10].cb(gpio_config[10].arg);
                        ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1079:43: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[10].cb(gpio_config[10].arg);
                                           ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1083:24: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[11].cb(gpio_config[11].arg);
                        ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1083:43: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[11].cb(gpio_config[11].arg);
                                           ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1087:24: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[12].cb(gpio_config[12].arg);
                        ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1087:43: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[12].cb(gpio_config[12].arg);
                                           ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1091:24: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[13].cb(gpio_config[13].arg);
                        ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1091:43: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[13].cb(gpio_config[13].arg);
                                           ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1095:24: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[14].cb(gpio_config[14].arg);
                        ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1095:43: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[14].cb(gpio_config[14].arg);
                                           ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1099:24: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[15].cb(gpio_config[15].arg);
                        ^
/home/lo/RIOT/cpu/samd21/periph/gpio.c:1099:43: warning: array subscript is above array bounds [-Warray-bounds]
             gpio_config[15].cb(gpio_config[15].arg);
                                           ^
"make" -C /home/lo/RIOT/drivers
"make" -C /home/lo/RIOT/sys
"make" -C /home/lo/RIOT/sys/auto_init
"make" -C /home/lo/RIOT/sys/newlib
/home/lo/RIOT/examples/hello-world/bin/samr21-xpro/core.a(hwtimer.o): In function `_hwtimer_set':
/home/lo/RIOT/core/hwtimer.c:169: undefined reference to `hwtimer_arch_set_absolute'
/home/lo/RIOT/core/hwtimer.c:173: undefined reference to `hwtimer_arch_set'
/home/lo/RIOT/examples/hello-world/bin/samr21-xpro/core.a(hwtimer.o): In function `hwtimer_spin':
/home/lo/RIOT/core/hwtimer.c:69: undefined reference to `hwtimer_arch_now'
/home/lo/RIOT/core/hwtimer.c:78: undefined reference to `hwtimer_arch_now'
/home/lo/RIOT/core/hwtimer.c:82: undefined reference to `hwtimer_arch_now'
/home/lo/RIOT/examples/hello-world/bin/samr21-xpro/core.a(hwtimer.o): In function `hwtimer_init_comp':
/home/lo/RIOT/core/hwtimer.c:96: undefined reference to `hwtimer_arch_init'
/home/lo/RIOT/examples/hello-world/bin/samr21-xpro/core.a(hwtimer.o): In function `hwtimer_now':
/home/lo/RIOT/core/hwtimer.c:116: undefined reference to `hwtimer_arch_now'
/home/lo/RIOT/examples/hello-world/bin/samr21-xpro/core.a(hwtimer.o): In function `hwtimer_remove':
/home/lo/RIOT/core/hwtimer.c:201: undefined reference to `hwtimer_arch_unset'
collect2: error: ld returned 1 exit status
/home/lo/RIOT/examples/hello-world/../../Makefile.include:181: recipe for target 'all' failed
make: *** [all] Error 1
```

**Comment by OlegHahm** [comment]
@LudwigOrtmann, scroll 3 comments up.

**Comment by LudwigOrtmann** [comment]
Did, saw it reported already, won't report again.

**Comment by kaspar030** [comment]
- rebased

**Comment by kaspar030** [comment]
(the samd21 warnings *should* not cause a problem, as the mentioned array accesses can only be triggered by previously activating the corresponding exti line, but as they are deactivated in periph_conf.h, they can't be setup anyways...)

**Comment by kaspar030** [comment]
- ahja, fixed samr21 compilation

**Comment by kaspar030** [comment]
Anyone? I'm tempted to remove the commits of those boards that are not tested, yet, and open them as seperate PRs.

**Comment by OlegHahm** [comment]
Currently, I can not even build for samr21-xpro:
`/home/oleg/git/RIOT/cpu/Makefile.include.cortex_common:34: /home/oleg/git/RIOT/cpu/_common/Makefile.include: No such file or directory`

**Comment by kaspar030** [comment]
@OlegHahm Sorry, works now.

**Comment by OlegHahm** [comment]
With this patch samr21 hangs when using `txtsnd` in default example.

**Comment by kaspar030** [comment]
@OlegHahm Does that work in master?

**Comment by OlegHahm** [comment]
Yes, it does.

**Comment by kaspar030** [comment]
@OlegHahm damn, and I don't have the board... If you'd like to debug this, could you try to set optimization back to -O0 in ```boards/Makefile.include.cortex_common``` and see if that fixes the problem? Also I've revised #2865, that should theoretially solve the problem, too.

**Comment by OlegHahm** [comment]
> If you'd like to debug this, could you try to set optimization back to -O0 in boards/Makefile.include.cortex_common and see if that fixes the problem?

Merging #2865 with `-Os` doesn't seem to work, either. Should I try to give you remote access?

**Comment by kaspar030** [comment]
@OlegHahm OK, I've sent you my ssh public key...

**Comment by haukepetersen** [comment]
tested the `msbiot` and the `stm32f0discovery`, both fine.

Regarding the `samr21-xpro`: As far as I see, this PR works, but the board itself does not cope with `-Os`. Besides the GPIO warnings (that are fixed in #2865), the radio driver initialization fails as soon as you switch from `-O0` to `-Os`. My guess that there are more problems with GPIO/SPI periph drivers somewhere...

**Comment by kaspar030** [comment]
Yeah, I spent some time debugging this. I ended up porting over the SPI and GPIO implementation from SAM L21, but at some point I remotely bricked Oleg's SAM R21 and couldn't finish.
I think I'll change this PR so it's possible to override -Os.

**Comment by OlegHahm** [comment]
You bricked it?

**Comment by haukepetersen** [comment]
> I think I'll change this PR so it's possible to override -Os.

+1 for this solution, so we can merge this independently from debugging the R21...

**Comment by kaspar030** [comment]
done...

**Comment by kaspar030** [comment]
@OlegHahm It doesn't flash remotely. I guess either with the right timing or with atmel studio we can unbrick. Don't worry. ;)

**Comment by OlegHahm** [comment]
stopped worrying.

**Comment by haukepetersen** [comment]
> done...

looks good. Now works for the `samr21-xpro`.

**Comment by kaspar030** [comment]
- rebased

**Comment by kaspar030** [comment]
btw, I'm not updating the parent PR, so I'll close that and remove the label.

**Comment by kaspar030** [comment]
Three more to go:

- [ ] nucleo-l1 (stm32l1)
- [ ] openmote (cc2538)
- [ ] udoo (sam3x8e)

**Comment by OlegHahm** [comment]
I think all of these nodes should available at FU Berlin somewhere.

**Comment by haukepetersen** [comment]
I am just testing for the udoo, but I cannot flash it. But this problem is also exists for the current master. Maybe its my board here... For the other two, I will try to find them somewhere here.

**Comment by haukepetersen** [comment]
openmote: does not link: `cannot open linker script file cc2538_linkerscript.ld: No such file or directory`. I think you have to include the cpu path for the `INCLUDE` statement in the linkerscripts.

**Comment by haukepetersen** [comment]
not sure how we handle this, but I think the solution for the k60 looks best. So how about we include
```
export LINKFLAGS += -L$(RIOTCPU)/$(CPU)
```
into `cpu/Makefile.include.cortex_common` or into the `board/Makefile.include.cortex_common`?

**Comment by haukepetersen** [comment]
'nucleo-l1': works.

**Comment by kaspar030** [comment]
I've added the path to ```board/Makefile.include.cortex_common``` as the other LINKERFLAGS are there, too.

**Comment by haukepetersen** [comment]
nice, `openmote` works

**Comment by haukepetersen** [comment]
alright, `udoo` does also work, took a while to get it to flash, but that was unrelated to this PR...

**Comment by kaspar030** [comment]
alright, then I squash!

**Comment by haukepetersen** [comment]
ACK when squashed an Travis is happy.

**Comment by kaspar030** [comment]
I kept the cpu/board Makefiles split up for easier reverting in case something goes wrong. That alright?

**Comment by OlegHahm** [comment]
:+1:

**Comment by kaspar030** [comment]
... and rebased again... travis, go go go! ;)

**Comment by haukepetersen** [comment]
Travis is complaining about whitespace errors...

**Comment by kaspar030** [comment]
- rebased
- fixed newline errors

**Comment by haukepetersen** [comment]
go Travis, go!

**Comment by kaspar030** [comment]
- squashrefactored

**Comment by haukepetersen** [comment]
@kaspar030: the `test/pkt_micro-ecc` does not compile for the `stm32f4discovery`. From a quick look, this seems to be related to the FPU. Seems to me, that the micro-eec package is build without FPU support, while the board is build with FPU support. Care to investigate?

**Comment by kaspar030** [comment]
IMHO that's essentially a bug in the make system... See #3046.

**Comment by haukepetersen** [comment]
be it or not, it needs to be solved (soon) before we can merge this PR, right?!

**Comment by kaspar030** [comment]
I think I got it. The CFLAGS_FPU where never actually evaluated as the board makefile was included before the cpu makefile, but board would only set the option if the CPU has set CORTEX before. I moved the check into the common CPU makefile. Let's see what travis says.

**Comment by haukepetersen** [comment]
ACK and go.

**Pull request merged**